src/rdkafka_cgrp.c

```diff
@@ -2666,21 +2692,76 @@ static rd_kafka_op_res_t rd_kafka_cgrp_consumer_handle_next_assignment(
         return RD_KAFKA_OP_RES_HANDLED;
 }
 
+static rd_kafka_topic_partition_list_t *
+rd_kafka_cgrp_consumer_assignment_with_metadata(
+    rd_kafka_cgrp_t *rkcg,
+    rd_kafka_topic_partition_list_t *assignment,
+    rd_list_t **missing_topic_ids) {
+        int i;
+        rd_kafka_t *rk = rkcg->rkcg_rk;
+        rd_kafka_topic_partition_list_t *assignment_with_metadata =
+            rd_kafka_topic_partition_list_new(assignment->cnt);
+        for (i = 0; i < assignment->cnt; i++) {
+                struct rd_kafka_metadata_cache_entry *rkmce;
+                rd_kafka_topic_partition_t *rktpar;
+                rd_kafka_Uuid_t request_topic_id =
+                    rd_kafka_topic_partition_get_topic_id(
+                        &assignment->elems[i]);
+                rd_kafka_rdlock(rk);
+                rkmce =
+                    rd_kafka_metadata_cache_find_by_id(rk, request_topic_id, 1);
+
+                if (rkmce) {
+                        rd_kafka_topic_partition_list_add_with_topic_name_and_id(
+                            assignment_with_metadata, request_topic_id,
+                            rkmce->rkmce_mtopic.topic,
+                            assignment->elems[i].partition);
+                        rd_kafka_rdunlock(rk);
+                        continue;
+                }
+                rd_kafka_rdunlock(rk);
+
+                rktpar = rd_kafka_topic_partition_list_find_topic_by_id(
+                    rkcg->rkcg_current_assignment, request_topic_id);
+                if (rktpar) {
+                        rd_kafka_topic_partition_list_add_with_topic_name_and_id(
+                            assignment_with_metadata, request_topic_id,
+                            rktpar->topic, assignment->elems[i].partition);
+                        continue;
+                }
+
+                if (missing_topic_ids) {
+                        rd_kafka_Uuid_t topic_id;
+                        if (unlikely(!*missing_topic_ids))
+                                *missing_topic_ids =
+                                    rd_list_new(1, rd_list_Uuid_destroy);
+                        topic_id = rd_kafka_topic_partition_get_topic_id(
+                            &assignment->elems[i]);
+                        rd_list_add(*missing_topic_ids,
+                                    rd_kafka_Uuid_copy(&topic_id));
+                }
+                rd_kafka_dbg(rkcg->rkcg_rk, CGRP, "HEARTBEAT",
+                             "Metadata not found for the "
+                             "assigned topic id - %s."
+                             " Continuing without it",
+                             rd_kafka_Uuid_base64str(&request_topic_id));
+        }
+        if (missing_topic_ids && *missing_topic_ids)
+                rd_list_deduplicate(missing_topic_ids,
+                                    (void *)rd_kafka_Uuid_ptr_cmp);
```

Comment:
Add a FIXME to change the name of `rd_list_deduplicate` as it sorts as well. Ideally, the function should not change the order. If we don't add sorted then it might be assumed that the order is not changed.

Reply:
I can add it to the documentation of this function later, the `missing_topic_ids` is just created so it doesn't change the order of the assignment. It's just used for the Metadata request